### PR TITLE
Decouple the discovery and refactor the retrieval package

### PR DESCRIPTION
**Pull request by krasi-georgiev**
now discovery it completely decoupled and uses a channel to send updates to the scraping loops.

TODO
--------------------
- [x] fix the  tests
- [x] fix the  comments
- [x] update the readme

Next PRs
---------------------
- Move Static Provider into sub-package like the other discovery providers
- Rename package retrieval to scrape
- Make all discovery providers private since we need them only from inside the Discovery package now

### CODE REVIEW

File: discovery/manager.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
+// Copyright 2016 The Prometheus Authors
```
Comment (brancz) [inline discovery/manager.go]: could you do a git mv here, otherwise it's very hard to follow the history of this file

Comment (krasi-georgiev) [reply discovery/manager.go]: good point. I rearranged the commits and used git mv ,
Now when you click on the actual commits it shows it properly , 
https://github.com/prometheus/prometheus/pull/3362/commits/8b656714002bd01ae544cdd23453656f85d64c99 , but the final PR still shows as if the file is deleted and a new one is created.


the PR is just  to get some opinions and than I will arrange it properly and add some tests as well.

File: retrieval/manager.go [hunk 2]
```diff
@@ -14,159 +14,133 @@
...
+//
+func NewScrapeManager(ctx context.Context, logger log.Logger, app Appendable) *ScrapeManager {
+
+	return &ScrapeManager{
+		ctx:           ctx,
+		append:        app,
+		logger:        logger,
+		actionCh:      make(chan func()),
+		scrapeConfigs: make(map[string]*config.ScrapeConfig),
+		scrapePools: make(map[string]struct {
```
Comment (brancz) [inline retrieval/manager.go]: why not make this a proper struct definition?

File: retrieval/manager.go [hunk 3]
```diff
@@ -14,159 +14,133 @@
...
+			scrapePool *scrapePool
+			ctx        context.Context
```
Comment (brancz) [inline retrieval/manager.go]: I'm having a hard time distinguishing the lifetime of this ctx vs scrapePool.ctx. I believe it's not needed here. Then we can just have the type be `map[string]*scrapePool`.

Comment (krasi-georgiev) [reply retrieval/manager.go]: thanks that simplifies it, I bit of confusion on my side.

**Comment by brancz** [comment]
This generally lgtm, although someone should do another review, as this is a pretty fragile piece. Maybe after fixing tests and adding comments though.

**Comment by krasi-georgiev** [comment]
thasnk @brancz 
@fabxc it is your turn now :)
Once confirmed all good , will fix the tests and the comments

**Comment by cstyan** [comment]
I'm still planning to do a review but haven't been able to get to it yet

File: discovery/manager.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+	if m.targetSetProviders[tsName] == nil {
+		m.targetSetProviders[tsName] = make(map[string]*targetSetProvider)
+	}
+	m.targetSetProviders[tsName][provName] = &targetSetProvider{
+		cancel:  cancel,
+		tgroups: []*config.TargetGroup{},
+	}
+}
+
+// mergeGroups adds a new target group for a named discovery provider and returns all  target groups for a given target set
```
Comment (Conorbro) [inline discovery/manager.go]: Nitpick: Double space in comment here. Comments should also end with a period.

File: cmd/prometheus/main.go [hunk 5]
```diff
@@ -288,6 +292,17 @@ func main() {
 
 	var g group.Group
 	{
+		g.Add(
+			func() error {
+				return discoveryManager.Run()
+			},
+			func(err error) {
+				level.Info(logger).Log("msg", "Stopping target manager...")
```
Comment (Conorbro) [inline cmd/prometheus/main.go]: Should this be "discovery manager" rather than "target manager" ?

File: discovery/manager.go (hunk 4 above)
Comment (krasi-georgiev) [reply discovery/manager.go]: thanks 
I will fix all comments and tests after I get one more OK from the core maintainers as it might require more changes and want to preserve some keystrokes :)

File: cmd/prometheus/main.go (hunk 5 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: thanks. noted will change that when I do the comments

**Comment by krasi-georgiev** [comment]
@Conorbro , do you see  anything wrong with this refactoring in general?

**Comment by Conorbro** [comment]
@krasi-georgiev As far as I can tell it seems fine although I would wait for the thumbs up from someone more familiar with this package. 

I also tested with ec2 sd and it all worked as expected.

File: cmd/prometheus/main.go [hunk 6]
```diff
@@ -288,6 +292,17 @@ func main() {
...
+				cancelCtx()
```
Comment (fabxc) [inline cmd/prometheus/main.go]: A single actor canceling a context used across components doesn't seem right. Each component should probably have it's own context with a dedicated cancel function.

File: cmd/prometheus/main.go [hunk 7]
```diff
@@ -288,6 +292,17 @@ func main() {
...
+				level.Info(logger).Log("msg", "Target manager stopped")
```
Comment (fabxc) [inline cmd/prometheus/main.go]: I think those "stopped" logs should be coming after the run function terminated.
Like this, both log lines will be emitted right after each other. That does not tell us whether the component might actually be hanging on shutdown.

File: cmd/prometheus/main.go [hunk 8]
```diff
@@ -450,17 +465,14 @@ func main() {
...
-				<-cancel
-				return nil
+				return scrapeManager.Run(discoveryManager.SyncCh())
 			},
 			func(err error) {
-				targetManager.Stop()
-				close(cancel)
+				level.Info(logger).Log("msg", "Stopping scrape manager...")
+				cancelCtx()
+				level.Info(logger).Log("msg", "Scrape manager stopped")
```
Comment (fabxc) [inline cmd/prometheus/main.go]: Same, this should be a defer in the run func above.

The first log line should probably also go there. At least from my understanding this function here should only send initial signal(s) for shutdown to the main runner func.

**Comment by krasi-georgiev** [comment]
@fabxc I did use git mv but for some reason when you have more than one commit it doesn't show the diff properly.
I will try again.

in the meanwhile if you click a single commit it shows it ok
https://github.com/prometheus/prometheus/pull/3362/commits/fec5e17840aac03918e276b94838db1c21c51227

File: cmd/prometheus/main.go (hunk 6 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: ok , will change.
would be interested if you can think of  an example when using a single context can cause an issue.

Comment (fabxc) [reply cmd/prometheus/main.go]: Cannot think of something for this particular case. But as soon as you inspect the interrupt error and decide to handle it specifically, any dependency of one actor on the interrupt function on another will get you in trouble I'd imagine.

Seems a good idea in general to avoid it all together.

File: cmd/prometheus/main.go (hunk 8 above)
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: yeas, I overlooked these parts a bit.
if you don't mind for such a small block I prefer that I do it without a `defer` for a more natural flow look.

```
func() error {
	err := scrapeManager.Run(discoveryManager.SyncCh())
	level.Info(logger).Log("msg", "Scrape manager stopped")
	return err
	},
func(err error) {
	level.Info(logger).Log("msg", "Stopping scrape manager...")
	cancelCtx()
	},
```

by the way I don't see how can we put the `Stopping scrape manager...` in the first block?
we need this log to show as soon as we start the cancelation and that can't happen with the blocking `scrapeManager.Run` function.

**Comment by krasi-georgiev** [comment]
@fabxc , @brancz - the problem with the git diff showing files as deleted is that the file changes are too severe. I couldn't find any way to set a different threshold so the only way to see the differences properly is locally by running `git diff -M10% HEAD~`

contacted github about the issue and if interested can update you with their reply.

I have rebased and now fixing the tests.

**Comment by fabxc** [comment]
So overall this looks sane to me. The entire area is historically rather fragile when changes are attempted. So this should ideally be tested in a reasonably complex environment for a bit.

One thing that concerns me is that the scrape manager seems to no longer have a graceful shutdown. How does this play with the general changes in main?
TSDB is not built to work with the "fire and forget" cancellation approach the context package has. Just canceling a context and tearing down everything without waiting for completion may leave you with scrape routines still appending data while the database is closed already. That will probably not end well.

**Comment by krasi-georgiev** [comment]
@fabxc I think your comment saved quite a few crashes :)
Now the `scrapeManager` doesn't rely on context for the cancelation.
Added a blocking `Stop()` function so  that `main.go` will run `fanoutStorage.Close()` only after `Stop()` exits - meaning all scrape loops have been stopped.

File: discovery/manager.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+		select {
+		case f := <-m.actionCh:
+			f()
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		}
+	}
+}
+
+// SyncCh returns a read only channel used by all DiscoveryProviders targetSet updates
```
Comment (cstyan) [inline discovery/manager.go]: it's a bit confusing that we no longer have `type TargetSet` but we're still referring to that type in comments, I know this is just Discoverer and Manager now, but can we find a way to reword the comments a bit?

File: discovery/manager.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+func (m *Manager) SyncCh() <-chan map[string][]*config.TargetGroup {
+	return m.syncCh
+}
+
+// ApplyConfig removes all running discovery providers and starts new ones using the provided config.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	err := make(chan error)
+	m.actionCh <- func() {
+		m.cancelDiscoverers()
```
Comment (cstyan) [inline discovery/manager.go]: should we not be checking if `m.cancelDiscoverers` exists as we were before?

File: retrieval/manager.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		actionCh:      make(chan func()),
+		scrapeConfigs: make(map[string]*config.ScrapeConfig),
+		scrapePools:   make(map[string]*scrapePool),
+		graceShut:     make(chan struct{}),
+	}
+}
+
+// ScrapeManager maintains a set of scrape pools and manages start/stop cicles
+// when receiving new target groups form the discovery manager.
+type ScrapeManager struct {
```
Comment (cstyan) [inline retrieval/manager.go]: Not sure if this makes sense, but it feels to me like the discovery package should provide an interface that other packages can implement to handle recv'ing  of target groups and then implementation details of how the consumer wants to use those groups is all the other package needs to do. Maybe that's actually what we're doing and I'm misreading?

File: retrieval/manager.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+		for _, scfg := range cfg.ScrapeConfigs {
+			m.scrapeConfigs[scfg.JobName] = scfg
+		}
+		close(done)
+	}
+	<-done
+	return nil
+}
+
+// Targets returns the targets currently being scraped.
```
Comment (cstyan) [inline retrieval/manager.go]: From this comment I think I'm not understanding what `scrapePool`s actually are. Should `Targets` return all the targets that we're currently scraping at some moment in time, or all targets that the `ScrapeManager` knows about?

File: discovery/manager.go (hunk 9 above)
Comment (krasi-georgiev) [reply discovery/manager.go]: thanks , I updated the comment

File: discovery/manager.go (hunk 10 above)
Comment (krasi-georgiev) [reply discovery/manager.go]: not sure what you mean , how can not exist?

File: retrieval/manager.go (hunk 11 above)
Comment (krasi-georgiev) [reply retrieval/manager.go]: can you give an example of maybe a simple gist with your idea.

File: retrieval/manager.go (hunk 12 above)
Comment (krasi-georgiev) [reply retrieval/manager.go]: targets that are currently being scrapped. this function is only used by the web gui - `/targets`

File: retrieval/manager.go (hunk 11 above)
Comment (cstyan) [reply retrieval/manager.go]: sure, I'll give it a shot, or perhaps we can discuss on irc?

File: discovery/manager.go (hunk 10 above)
Comment (cstyan) [reply discovery/manager.go]: Sorry, I guess should we be checking if the cancel func for `m.discoverCancel` exists, misread how this had changed. See my new comments below

File: discovery/manager.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+	}
+
+	return <-err
+}
+
+func (m *Manager) startProvider(jobName, provName string, worker Discoverer) {
+	ctx, cancel := context.WithCancel(m.ctx)
+	updates := make(chan []*config.TargetGroup)
+
+	m.discoverCancel = append(m.discoverCancel, cancel)
```
Comment (cstyan) [inline discovery/manager.go]: so here we add the cancel functions

Comment (cstyan) [reply discovery/manager.go]: and `startProvider` is called later in `ApplyConfig` than `m.cancelDiscoverers`

File: discovery/manager.go [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+					return
+				}
+				m.syncCh <- m.mergeGroups(jobName, provName, tgs)
+			}
+		}
+	}(provName)
+}
+
+func (m *Manager) cancelDiscoverers() {
+	for _, c := range m.discoverCancel {
```
Comment (cstyan) [inline discovery/manager.go]: so what if `c` has not been set via `startProvider` yet?

File: discovery/manager.go (hunk 10 above)
Comment (krasi-georgiev) [reply discovery/manager.go]: the `NewManager` constructor initialises to a zero value so no need for an extra check.

File: retrieval/manager.go (hunk 11 above)
Comment (krasi-georgiev) [reply retrieval/manager.go]: sure , fork it  and change it to your idea and can discuss.

File: discovery/manager.go (hunk 10 above)
Comment (cstyan) [reply discovery/manager.go]: facepalming, missed that `discoverCancel` was a field in the `Manager` struct

thanks! :)

**Comment by krasi-georgiev** [comment]
@fabxc @brancz  this is 100% ready. 
Is there any easy way to test for more complex setups?
I tried locally with `consul` and @Conorbro confirmed for `ec2`, but like you mentioned it would be best to test in something different than local.

btw the failure is the flaky file sd test which I give 93.54% gurantee to be fixed in
https://github.com/prometheus/prometheus/pull/3492 💣
😄

**Comment by fabxc** [comment]
Any chance you can get that tested in OpenShift/Kubernetes?

File: discovery/manager.go [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+			case tgs, ok := <-updates:
+				// Handle the case that a target provider exits and closes the channel
+				// before the context is done.
+				if !ok {
+					return
+				}
+				m.syncCh <- m.mergeGroups(jobName, provName, tgs)
+			}
+		}
+	}(provName)
```
Comment (fabxc) [inline discovery/manager.go]: This is not in a loop so there's no reason to pass this as an argument.

File: retrieval/manager.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		append:        app,
+		logger:        logger,
+		actionCh:      make(chan func()),
+		scrapeConfigs: make(map[string]*config.ScrapeConfig),
+		scrapePools:   make(map[string]*scrapePool),
+		graceShut:     make(chan struct{}),
+	}
+}
+
+// ScrapeManager maintains a set of scrape pools and manages start/stop cicles
```
Comment (fabxc) [inline retrieval/manager.go]: cycles

File: retrieval/scrape.go [hunk 17]
```diff
@@ -151,14 +150,15 @@ func newScrapePool(ctx context.Context, cfg *config.ScrapeConfig, app Appendable
 	sp := &scrapePool{
 		appendable: app,
 		config:     cfg,
-		ctx:        ctx,
```
Comment (fabxc) [inline retrieval/scrape.go]: Wait, this was never used/canceled? But it was the context passed to `newScrapeLoop` before which holds it and actually checks it in multiple places for cancelation. If that path was never possible, we'd have a logical flaw somehow, no?

File: discovery/manager.go [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+	Run(ctx context.Context, up chan<- []*config.TargetGroup)
+}
+
+// type pool struct {
+// 	cancel func()
+// 	tgroups []*config.TargetGroup
+// }
+
+// NewManager is the Discovery Manager constructor
+func NewManager(ctx context.Context, logger log.Logger) *Manager {
```
Comment (fabxc) [inline discovery/manager.go]: The constructor itself does no long-running processing that listens on the context. So maybe this one would be better placed as an argument to `Run()`? By extension the action functions would then receive a context arg which `Run()` injects.

File: discovery/manager.go [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+type Manager struct {
+	ctx            context.Context
+	logger         log.Logger
+	syncCh         chan map[string][]*config.TargetGroup // map[targetSetName]
+	actionCh       chan func()
+	discoverCancel []context.CancelFunc
+	targets        map[string]map[string][]*config.TargetGroup // map[targetSetName]map[providerName]
```
Comment (fabxc) [inline discovery/manager.go]: A full comment might be better than the custom syntax style there.

File: discovery/manager.go [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+		}
+	}(provName)
+}
+
+func (m *Manager) cancelDiscoverers() {
+	for _, c := range m.discoverCancel {
+		c()
+	}
+	m.targets = make(map[string]map[string][]*config.TargetGroup)
+	m.discoverCancel = []context.CancelFunc{}
```
Comment (fabxc) [inline discovery/manager.go]: Either just set it to `nil` or, if you are allocation savvy to `m.discoverCancel[:0]`. Explicitly instantiating an empty slice with no capacity pre-allocation never really adds anything.

File: discovery/manager.go (hunk 15 above)
Comment (fabxc) [reply discovery/manager.go]: I'd in fact rename the function to `runProvider` and have it block within that for loop. Then use `go m.runProvider(...)` at the call site.
Starting  a producer go routine is fine and obviously necessary but completely detaching work of a method often just makes it harder to reuse in other contexts (mostly testing).

File: discovery/manager.go [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+
+// mergeGroups adds a new target group for a given discovery provider and returns all target groups for a given target set
+func (m *Manager) mergeGroups(tsName, provName string, tg []*config.TargetGroup) map[string][]*config.TargetGroup {
+	tset := make(chan map[string][]*config.TargetGroup)
+
+	m.actionCh <- func() {
+		if m.targets[tsName] == nil {
+			m.targets[tsName] = make(map[string][]*config.TargetGroup)
+		}
+		m.targets[tsName][provName] = []*config.TargetGroup{}
```
Comment (fabxc) [inline discovery/manager.go]: Same as above. no need to instantiate an empty slice unless for pre-allocation. This line can just be dropped.

File: discovery/manager.go [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+
+		if tg != nil {
+			m.targets[tsName][provName] = tg
+		}
+		tgAll := []*config.TargetGroup{}
```
Comment (fabxc) [inline discovery/manager.go]: `var tgAll []*config.TargetGroup`

File: discovery/manager.go [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+					tgAll = append(tgAll, tg)
+				}
+			}
+		}
+		t := make(map[string][]*config.TargetGroup)
+		t[tsName] = tgAll
+		tset <- t
+	}
+	return <-tset
+}
```
Comment (fabxc) [inline discovery/manager.go]: I found this a bit hard to follow. It seems to add a target group for a specific provider. But the second half is all about getting data from all providers.
Would it maybe be more reasonable to split this into distinct functions and above first call `m.addGroup(tsName, provName, group)` and then `m.syncCh <- m.allGroups(tsName)`?

File: discovery/manager_test.go [hunk 24]
```diff
@@ -0,0 +1,726 @@
...
+
+func (tp mockdiscoveryProvider) Run(ctx context.Context, up chan<- []*config.TargetGroup) {
+	tp.up = up
+	tp.sendUpdates()
+}
+
+func (tp mockdiscoveryProvider) sendUpdates() {
+	for _, update := range tp.updates {
+
+		time.Sleep(update.interval * time.Millisecond)
```
Comment (fabxc) [inline discovery/manager_test.go]: Where do you suspect the code to be timing dependent to have the tests play with sleeps?
Also `update.interval` already is a duration but the test cases just add plain numbers, which would be nanoseconds. A proper duration should be specified there rather than implicitly multiplying it with milliseconds further down.

**Comment by simonpasquier** [comment]
@krasi-georgiev @fabxc I did a quick test of the patch on OpenShift 3.7 and it's working fine. I'll do more in-depth testing today.

File: cmd/prometheus/main.go [hunk 25]
```diff
@@ -229,26 +230,29 @@ func main() {
 
 	cfg.queryEngine.Logger = log.With(logger, "component", "query engine")
 	var (
-		notifier       = notifier.New(&cfg.notifier, log.With(logger, "component", "notifier"))
-		targetManager  = retrieval.NewTargetManager(fanoutStorage, log.With(logger, "component", "target manager"))
-		queryEngine    = promql.NewEngine(fanoutStorage, &cfg.queryEngine)
-		ctx, cancelCtx = context.WithCancel(context.Background())
+		ctxWeb, cancelWeb             = context.WithCancel(context.Background())
+		ctxDiscovery, cancelDiscovery = context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply cmd/prometheus/main.go]: Yes I prefer as well but they are they are used just after the definition so can't move them

**Comment by krasi-georgiev** [comment]
@simonpasquier thanks, that would really help!

File: discovery/manager.go [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		targets:        make(map[string]map[string][]*config.TargetGroup),
+		discoverCancel: []context.CancelFunc{},
+	}
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
+// Targets pool is kept in a map with a format map[targetSetName]map[providerName].
+type Manager struct {
```
Comment (fabxc) [inline discovery/manager.go]: The docs on the internal maps should be moved above the fields. Otherwise they will show those implementation details in the regular Go doc page.s

File: discovery/manager.go [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+// 	tgroups []*config.TargetGroup
+// }
+
+// NewManager is the Discovery Manager constructor
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger:         logger,
+		actionCh:       make(chan func(context.Context)),
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		targets:        make(map[string]map[string][]*config.TargetGroup),
```
Comment (fabxc) [inline discovery/manager.go]: One thing I didn't notice last time, can we move this to a one-dimensional map like:

```go
type poolKey struct {
    set      string
    provider string
}

map[poolKey][]*config.TargetGroup
```

Or does that make things more complex elsewhere?

Comment (krasi-georgiev) [reply discovery/manager.go]: @fabxc  I completely forgot you can do that. This change makes is much more readable. Thanks!

also updated the comment.

File: discovery/manager_test.go (hunk 24 above)
Comment (krasi-georgiev) [reply discovery/manager_test.go]: Without sleeps you can't predict the order of the updates so you can't match with the expected result for the intermediate updates.  
The original discovery test was like this so I thought it covered some bug for the intermediate updates. If that is not the case than I can remove the sleeps and just match the final result to the expected result.

File: retrieval/scrape.go (hunk 17 above)
Comment (krasi-georgiev) [reply retrieval/scrape.go]: yes you are right it is referenced in the `mainLoop:` so this was never reachable so I added another commit
```
select {
		case <-sl.ctx.Done():
			close(sl.stopped)
			return
		case <-sl.scrapeCtx.Done():
			break mainLoop
		case <-ticker.C:
		}
```
I think the Scrape Loops logic needs a bit of refactoring as well, but  I will do that in another PR.

**Comment by krasi-georgiev** [comment]
asked for some testing in these SD issues so should have some more feedback soon.
https://github.com/prometheus/prometheus/issues/3007#issuecomment-348632689
https://github.com/prometheus/prometheus/issues/2335#issuecomment-348830720

**Comment by krasi-georgiev** [comment]
here is a Linux amd64 binary for anyone willing to test it before the merge
https://github.com/krasi-georgiev/prometheus/releases/tag/v2.0.0-beta.x

**Comment by fabxc** [comment]
Since this has seen a few rounds of review and real-world test, I'm merging this now. By its nature it's a huge change and there's no 100% certainty but its best to get this out into the wild now :)

Thanks for your effort and patience!

**Comment by krasi-georgiev** [comment]
@brian-brazil this also closes https://github.com/prometheus/prometheus/pull/3362
